security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -582,9 +582,25 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
 
 		// Re-generate secret if it's expired.
 		if sc.shouldRotate(&secret) {
+			// withToken indicates whether we use the token in our CSR request
+			withToken := true
 			atomic.AddUint64(&sc.secretChangedCount, 1)
-			// Send the notification to close the stream if token is expired, so that client could re-connect with a new token.
-			if sc.isTokenExpired(&secret) {
+
+			// if user provides a PROV_CERT Directory Path for client to extract path
+			// we will use cert to provide authentication and not use a token in that case
+			// and reset the connection to use the previously rotated cert to do mtls
+			// otherwise we will first check whether the token is expired or not
+			if sc.secOpts.ProvCert != "" {
+				withToken = false
+				err := sc.fetcher.ReconnectCaClient.Reconnect()
```

Comment:
Maybe I'm missing some context, could you clarify a little bit why we need to reconnect in this case? As long as `ProvCert != ""`, doesn't `NewCitadelClient()` already checks the `ProvCert != ""` in the first place which means the connection is already using the certificate?

Reply:
if it the ProvCert is provided,  means that the client should use this place to extract cert (User should set the OUTPUT_CERTS to the same as the ProvCert). in this case we should not use token anymore